fix(cmake): fix msvc building error

zylalx1 opened the pull request:
fix #6931 #7371 
(similar issue exists when using msvc as compiler, and ninja or nmake as generator, they are also being fixed. )

further investigation of the issue:

CMake generates a PreLinkEvent in the "lvgl.vcxproj" when using visual studio as the generator. 

```
    <PreLinkEvent>
      <UseUtf8Encoding>Always</UseUtf8Encoding>
      <Message>Auto build dll exports</Message>
      <Command>setlocal
cd D:\lvgl_windows_9.3\build
if %errorlevel% neq 0 goto :cmEnd
D:
if %errorlevel% neq 0 goto :cmEnd
"C:\Program Files\CMake\bin\cmake.exe" -E __create_def D:/lvgl_windows_9.3/build/lvgl.dir/Debug/exports.def D:/lvgl_windows_9.3/build/lvgl.dir/Debug//objects.txt
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal &amp; call :cmErrorLevel %errorlevel% &amp; goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd</Command>
    </PreLinkEvent>
```

this line
`"C:\Program Files\CMake\bin\cmake.exe" -E __create_def D:/lvgl_windows_9.3/build/lvgl.dir/Debug/exports.def D:/lvgl_windows_9.3/build/lvgl.dir/Debug//objects.txt`
tries to generate the "export.def" file from all the object files listed in "objects.txt", including "lv_blend_helium.obj" and "lv_blend_neon.obj". but those two files do not exist so it fails, causing the error.


so here is the question: (0) should those two object files be included in the prerequisites of the dll library? (in fact they are useless on non-ARM platforms)

if the answer is no, (1) why are those files being included by CMake?
if the answer is yes, (2) why are those files not being generated by msvc?

after some research i found the answer of (1) is this part of code in "lvgl/CMakeLists.txt" doesn't work as expected:

```
if(NOT ESP_PLATFORM)
  if(NOT (CMAKE_C_COMPILER_ID STREQUAL "MSVC"))
    project(lvgl LANGUAGES C CXX ASM HOMEPAGE_URL https://github.com/lvgl/lvgl)
  else()
    project(lvgl LANGUAGES C CXX HOMEPAGE_URL https://github.com/lvgl/lvgl)
  endif()
endif()
```

the CMAKE_C_COMPILER_ID is empty before the C language being specified using project() or enable_language(), so the first branch with ASM is always taken, regardless of the compiler is msvc or not.
this PR fixes the bug. asm files will be excluded from "objects.txt" and the project should be able to build without error now. 

for (2), the reason is that the msvc itself (cl.exe) supports only inline assembly (__asm blocks) but not individual assembly files. for individual assembly files masm (ml64.exe) is required. 

masm support can be enabled this way:

in lvgl/CMakeLists.txt:
```
if(CMAKE_C_COMPILER_ID STREQUAL "MSVC")
  enable_language(ASM_MASM)
endif()
```

in lvgl/env_support/cmake/custom.cmake:
```
if(CMAKE_C_COMPILER_ID STREQUAL "MSVC")
  file(GLOB_RECURSE ASM_SOURCES "${LVGL_ROOT_DIR}/src/*.S")
  set_source_files_properties(${ASM_SOURCES} PROPERTIES LANGUAGE ASM_MASM)
endif()
```

by this modification masm will try to assemble those .S files. but unfortunately masm itself (ml64.exe) doesn't support C-style comments and preprocessing directives, also it won't call the cl.exe to do the preprocessing job, so it fails on the first line (/**):
```
  1>Checking Build System
  Building Custom Rule D:/lvgl_windows_9.3/lvgl/CMakeLists.txt
  Assembling D:\lvgl_windows_9.3\lvgl\src\draw\sw\blend\helium\lv_blend_helium.S...
D:\lvgl_windows_9.3\lvgl\src\draw\sw\blend\helium\lv_blend_helium.S(1): error A2008: syntax error : / [D:\lvgl_windows_9.3\build\lvgl.vcxproj]
...
```

to make things right we need to use customized build process on those files: call the preprocessor first (cl.exe /P) and then feed the output to the masm. there should be a way to do that in CMake but i haven't figured out by now. 

another option is to change those .S files to .c files with inline assembly that cl.exe supports. C-style comments and preprocessing directives in inline assembly are supported too. but i don't think it's worth taking the risk of making such big change. 

the simplest way may be better: just exclude those files from msvc build since they are very unlikely to be enabled on windows PCs. (microsoft said they have added support for ARM arch, but i've never tried to compile for ARM arch using msvc and masm on windows so far, don't know if it's possible)

kisvegabor reviewed (approved):
Thank you! 

cc @vwheeler63 @kdschlosser @MouriNaruto

liamHowatt commented:
I switched over to Windows to test. I am not seeing the same issue as you where the compiler ID is not set before the project is set. I also tried removing the `project` from my project's CMakeLists.txt txt but could not reproduce. Is your `CMAKE_C_COMPILER_ID` set to some other unexpected value maybe?

I was told `if (MSVC)` is a better way to check for Windows. It's used at the bottom of the file you changed. Maybe give that a try to see if it makes a difference.

In any case, I think this change is fine. Thanks.

liamHowatt commented:
I should mention, I tested on Windows 10 so maybe it's a difference with Windows 11.

zylalx1 commented:
> I switched over to Windows to test. I am not seeing the same issue as you where the compiler ID is not set before the project is set. I also tried removing the `project` from my project's CMakeLists.txt txt but could not reproduce. Is your `CMAKE_C_COMPILER_ID` set to some other unexpected value maybe?
> 
> I was told `if (MSVC)` is a better way to check for Windows. It's used at the bottom of the file you changed. Maybe give that a try to see if it makes a difference.
> 
> In any case, I think this change is fine. Thanks.

that sounds strange. i added some message() and did the test on a windows 10 pc with cmake 3.31.1 and vs 2022:
```
message(STATUS "before project()")
message(STATUS "CMAKE_GENERATOR: ${CMAKE_GENERATOR}")
message(STATUS "CMAKE_C_COMPILER_ID: ${CMAKE_C_COMPILER_ID}")
message(STATUS "CMAKE_CXX_COMPILER_ID: ${CMAKE_CXX_COMPILER_ID}")
message(STATUS "CMAKE_C_COMPILER_VERSION: ${CMAKE_C_COMPILER_VERSION}")
message(STATUS "CMAKE_CXX_COMPILER_VERSION: ${CMAKE_CXX_COMPILER_VERSION}")
if (MSVC)
  message(STATUS "MSVC IS TRUE")
else()
  message(STATUS "MSVC IS FALSE")
endif()

if(NOT ESP_PLATFORM)
  if(NOT (CMAKE_C_COMPILER_ID STREQUAL "MSVC"))
    message(STATUS "COMPILER IS NOT MSVC")
    project(lvgl LANGUAGES C CXX ASM HOMEPAGE_URL https://github.com/lvgl/lvgl)
  else()
    message(STATUS "COMPILER IS MSVC")
    project(lvgl LANGUAGES C CXX HOMEPAGE_URL https://github.com/lvgl/lvgl)
  endif()
endif()

message(STATUS "after project()")
message(STATUS "CMAKE_GENERATOR: ${CMAKE_GENERATOR}")
message(STATUS "CMAKE_C_COMPILER_ID: ${CMAKE_C_COMPILER_ID}")
message(STATUS "CMAKE_CXX_COMPILER_ID: ${CMAKE_CXX_COMPILER_ID}")
message(STATUS "CMAKE_C_COMPILER_VERSION: ${CMAKE_C_COMPILER_VERSION}")
message(STATUS "CMAKE_CXX_COMPILER_VERSION: ${CMAKE_CXX_COMPILER_VERSION}")

if (MSVC)
  message(STATUS "MSVC IS TRUE")
else()
  message(STATUS "MSVC IS FALSE")
endif()

if(NOT (CMAKE_C_COMPILER_ID STREQUAL "MSVC"))
  message(STATUS "COMPILER IS NOT MSVC")
else()
  message(STATUS "COMPILER IS MSVC")
endif()
```

with this command is excuted in developer command prompt for vs 2022:
`cmake --preset=windows-base`

and this is my result:
```
Preset CMake variables:

  BUILD_SHARED_LIBS="ON"
  CMAKE_CXX_COMPILER="cl.exe"
  CMAKE_C_COMPILER="cl.exe"
  CMAKE_INSTALL_PREFIX:PATH="D:/lvgl_windows/lvgl/build/install/windows-base"
  LV_CONF_SKIP:BOOL="TRUE"

-- before project()
-- CMAKE_GENERATOR: Visual Studio 17 2022
-- CMAKE_C_COMPILER_ID:
-- CMAKE_CXX_COMPILER_ID:
-- CMAKE_C_COMPILER_VERSION:
-- CMAKE_CXX_COMPILER_VERSION:
-- MSVC IS FALSE
-- COMPILER IS NOT MSVC
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.42.34435.0
-- The CXX compiler identification is MSVC 19.42.34435.0
-- The ASM compiler identification is MSVC
-- Found assembler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.42.34433/bin/Hostx64/x64/cl.exe
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.42.34433/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.42.34433/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- after project()
-- CMAKE_GENERATOR: Visual Studio 17 2022
-- CMAKE_C_COMPILER_ID: MSVC
-- CMAKE_CXX_COMPILER_ID: MSVC
-- CMAKE_C_COMPILER_VERSION: 19.42.34435.0
-- CMAKE_CXX_COMPILER_VERSION: 19.42.34435.0
-- MSVC IS TRUE
-- COMPILER IS MSVC
-- Configuring done (21.0s)
-- Generating done (0.6s)
-- Build files have been written to: D:/lvgl_windows/lvgl/build/windows-base
```

i guess there may still be some differences in the enviroment that we've overlooked?

kdschlosser commented:
I am not sure why MSVC is being set to false in cmake. It should be set to true if cl.exe is found.

anywho, here is some light reading on cmake and detecting MSVC

https://izzys.casa/2023/09/finding-msvc-with-cmake/

kdschlosser commented:
OK so this is what you need to do...


```
project(lvgl LANGUAGES C CXX HOMEPAGE_URL https://github.com/lvgl/lvgl)
if (MSVC)
  enable_language(ASM)
endif()

```

zylalx1 commented:
> I am not sure why MSVC is being set to false in cmake. It should be set to true if cl.exe is found.
> 
> anywho, here is some light reading on cmake and detecting MSVC
> 
> https://izzys.casa/2023/09/finding-msvc-with-cmake/

i think it is being set to true, but only after project(lvgl LANGUAGES C) or enable_language(C), at which point the compiler detection runs, like your code shows.

kdschlosser commented:
That is correct. which is the reason why you check for MSVC after and then you add the additional language then.
